framework/src/meshgenerators/MeshRepairGenerator.C

```diff
@@ -0,0 +1,162 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MeshRepairGenerator.h"
+#include "CastUniquePointer.h"
+#include "MooseMeshUtils.h"
+
+#include "libmesh/mesh_tools.h"
+#include "libmesh/mesh_modification.h"
+
+registerMooseObject("MooseApp", MeshRepairGenerator);
+
+InputParameters
+MeshRepairGenerator::validParams()
+{
+
+  InputParameters params = MeshGenerator::validParams();
+  params.addClassDescription(
+      "Mesh generator to perform various improvement / fixing operations on an input mesh");
+  params.addRequiredParam<MeshGeneratorName>("input",
+                                             "Name of the mesh generator providing the mesh");
+
+  params.addParam<bool>("fix_node_overlap", false, "Whether to merge overlapping nodes");
+  params.addParam<Real>("node_overlap_tol", 1e-8, "Tolerance for merging overlapping nodes");
+
+  params.addParam<bool>(
+      "fix_elements_orientation", false, "Whether to flip elements with negative volumes");
+
+  params.addParam<bool>("separate_blocks_by_element_types",
+                        false,
+                        "Create new blocks if multiple element types are present in a block");
+
+  return params;
+}
+
+MeshRepairGenerator::MeshRepairGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _input(getMesh("input")),
+    _fix_overlapping_nodes(getParam<bool>("fix_node_overlap")),
+    _node_overlap_tol(getParam<Real>("node_overlap_tol")),
+    _fix_element_orientation(getParam<bool>("fix_elements_orientation")),
+    _elem_type_separation(getParam<bool>("separate_blocks_by_element_types"))
+{
+}
+
+std::unique_ptr<MeshBase>
+MeshRepairGenerator::generate()
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+  mesh->prepare_for_use();
+
+  // Blanket ban on distributed. This can be relaxed for some operations if needed
+  if (!mesh->is_replicated())
+    mooseError("MeshRepairGenerator is not implemented for distributed meshes");
```

Comment:
is_serial feels replicated?
```
true if all elements and nodes of the mesh exist on the current processor, false otherwise
```

Reply:
we do replicated elsewhere

Reply:
The docs try to explain the intent of each method: "if all elements and nodes of the mesh exist on the current processor" for `is_serial()`, "if new elements and nodes can and should be created in synchronization on all processors" for `is_replicated()`.

If you're trying to figure out whether you can read any existing element on any processor, `is_serial()` is the right thing to use.  If you're trying to figure out whether you can create local elements independently from processor to processor, `!is_replicated()` is.  But IIRC the latter is an older API so I'm sure (even in libMesh) it's still being used in a ton of places where `is_serial()` would have been more appropriate.

Reply:
For mesh generators, I'm currently a fan of `MeshSerializer`, to replace the "are you serial?" question with "YOU ARE SERIAL", but allowing future steps in mesh generation to re-parallelize.

This still isn't quite right, though.  `MeshSerializer` both serializes and (in its destructor) unserializes.  Wasteful if the mesh is soon going to need to be serialized again ... but the alternative of serializing and leaving it serialized to the end is wasteful if the mesh is going to go through processes (many generators) that are more efficient on a distributed mesh.  I'm not sure what the efficient way to handle this is.  Have mesh generators register whether they need serialized inputs or not, then distribute an output mesh as soon as it's no longer going to be an input to a subsequent generator?

Reply:
It seems like a follow-on mesh generator that knows it's very heavy and can work with a distributed mesh, could check `!is_replicated() && is_serial()` and act accordingly. It's perhaps best for the current mesh generator to not guess, do less operations, register its state (accessible through APIs like `is_serial()` and `is_prepared()`) and assume follow-on generators will do what's best for themselves.

Reply:
Since this serial/replicated distinction paradigm has not made it to MOOSE, I can with either there but is_replicated seems like the more consistent treatment in MOOSE

Reply:
ok found it in a few places

Reply:
changed.